cylc/flow/scripts/profile.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""cylc profiler [OPTIONS]
+
+Profiler which periodically polls PBS cgroups to track
+the resource usage of jobs running on the node.
+"""
+
+import os
+import re
+import sys
+import time
+import signal
+import subprocess
+from pathlib import Path
+from dataclasses import dataclass
+from cylc.flow.terminal import cli_function
+from cylc.flow.option_parsers import CylcOptionParser as COP
+
+INTERNAL = True
+PID_REGEX = re.compile(r"([^:]*\d{6,}.*)")
+
+
+def get_option_parser() -> COP:
+    parser = COP(
+        __doc__,
+        argdoc=[
+        ],
+    )
+    parser.add_option(
+        "-i", type=int, help="interval between query cycles in seconds",
+        default=10, dest="delay")
+    parser.add_option(
+        "-m", type=str, help="Location of cgroups directory",
+        default="/sys/fs/cgroup",
+        dest="cgroup_location")
+
+    return parser
+
+
+@cli_function(get_option_parser)
+def main(parser, options):
+    """CLI main."""
+    # Register the stop_profiler function with the signal library
+    signal.signal(signal.SIGINT, stop_profiler)
```

Comment:
We might want to also register SIGTERM and SIGHUP.

Reply:
Done